### PR TITLE
generate durable JMS comsumer attributes for messaging adapter

### PR DESCRIPTION
**Pull request by jaceko**

### CODE REVIEW

File: adapters/messaging-adapter-generator/pom.xml [hunk 1]
```diff
@@ -39,6 +39,11 @@
             <version>${project.version}</version>
         </dependency>
         <dependency>
+            <groupId>org.antlr</groupId>
+            <artifactId>ST4</artifactId>
+            <version>4.0.8</version>
```
Comment (matt-rich) [inline adapters/messaging-adapter-generator/pom.xml]: Move version into parent

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java [hunk 2]
```diff
@@ -1,80 +1,80 @@
...
+     *                      source and
+     *                      destination folders
      */
     @Override
     public void run(final Raml raml, final GeneratorConfig configuration) {
 
         validator.validate(raml);
 
         final Collection<Resource> ramlResourceModels = raml.getResources().values();
+        BaseUri baseUri = new BaseUri(raml.getBaseUri());
```
Comment (matt-rich) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java]: make baseUri final

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java [hunk 3]
```diff
@@ -93,83 +93,83 @@ private String jmsListenerTemplate() {
...
+     * @param configuration Configuration information  @return template attributes
      */
-    private Attributes templateAttributesFrom(final Resource resource, final String baseUri, final GeneratorConfig configuration) {
-        final String resourceUri = resource.getUri();
-        final HashMap<String, String> data = new HashMap<>();
+    private TemplateAttributes templateAttributesFrom(final Resource resource, final BaseUri baseUri, final GeneratorConfig configuration) {
+        final ResourceUri resourceUri = new ResourceUri(resource.getUri());
+        final List<Pair<String, String>> mainAttributes = new LinkedList<>();
+        final List<Pair<String, String>> activationConfigAttributes = new LinkedList<>();
         Component component = componentOf(resourceUri, baseUri);
```
Comment (matt-rich) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java]: Make component final

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java [hunk 4]
```diff
@@ -93,83 +93,83 @@ private String jmsListenerTemplate() {
...
-        data.put("PACKAGE_NAME", configuration.getBasePackageName());
-        data.put("CLASS_NAME", classNameOf(resourceUri, baseUri));
-        data.put("ADAPTER_TYPE", component.name());
-        data.put("DESTINATION_TYPE", component.destinationType().getName());
-        data.put("DESTINATION_LOOKUP", destinationNameOf(resourceUri));
-        data.put("MESSAGE_SELECTOR", messageSelectorsFrom(resource.getActions()));
+        mainAttributes.add(Pair.of("PACKAGE_NAME", configuration.getBasePackageName()));
+        mainAttributes.add(Pair.of("CLASS_NAME", classNameOf(resourceUri, component)));
+        mainAttributes.add(Pair.of("ADAPTER_TYPE", component.name()));
+        activationConfigAttributes.add(Pair.of("destinationType", component.destinationType().getName()));
```
Comment (matt-rich) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java]: Make keys constants (final static) - e.g. "destinationType"

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java [hunk 5]
```diff
@@ -93,83 +93,83 @@ private String jmsListenerTemplate() {
...
      */
-    private Component componentOf(final String resourceUri, final String baseUri) {
-        final String[] uriParts = resourceUri.split("\\.");
-        String pillar = uriParts[uriParts.length - 1];
-        String tier = EVENTS.equals(pillar)
-                ? contains(baseUri, PROCESSOR) ? PROCESSOR : LISTENER
-                : uriParts[uriParts.length - 2];
+    private Component componentOf(final ResourceUri resourceUri, final BaseUri baseUri) {
+
+        String pillar = resourceUri.pillar();
```
Comment (matt-rich) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java]: Make pillar and tier final

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/uri/BaseUri.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.regex.Pattern;
+
+import static java.lang.String.format;
+
+/**
+ * Parses Raml base uri and exposes it's parts through accessor methods
+ */
+public class BaseUri {
+    private static final Pattern MESSAGING_BASE_URI_PATTERN
+            = Pattern.compile("message://(event|command|query)/(api|controller|handler|listener|processor)/\\S+/(\\S+)");
```
Comment (matt-rich) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/uri/BaseUri.java]: This needs to support 'view' in addition to api|controller|handler|listener|processor

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/uri/ResourceUri.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
+package uk.gov.justice.raml.jms.uri;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.apache.commons.lang3.StringUtils.stripStart;
+
+public class ResourceUri {
+    private static final Pattern MESSAGING_RESOURCE_URI_PATTERN =
+            Pattern.compile("^\\/?([a-zA-Z0-9]+)\\.(((api|controller|handler)\\.(commands))|(events))");
```
Comment (matt-rich) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/uri/ResourceUri.java]: The allowed values could be retrieved from the Component enum, with a non-null destination type.

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/uri/ResourceUri.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    private final String uriString;
+    private final String tier;
+    private final String context;
+    private final String pillar;
+
+    public ResourceUri(String uriString) {
+        this.uriString = uriString;
+        final Matcher m = matcherOf(this.uriString);
+        m.find();
+        this.context = m.group(1);
```
Comment (matt-rich) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/uri/ResourceUri.java]: Is it possible to extract the magic numbers to a meaninful named constant?

File: adapters/messaging-adapter-generator/src/test/java/uk/gov/justice/raml/jms/core/JmsEndpointGeneratorTest.java [hunk 9]
```diff
@@ -547,6 +549,51 @@ public void shouldCallJmsProcessorWhenOnMessageIsInvoked() throws Exception {
...
+                        .with(resource()
+                                .withRelativeUri("/people.events")
+                                .with(action(POST, "application/vnd.context1.events.abc+json")))
+                        .build(),
+                configurationWithBasePackage(BASE_PACKAGE, outputFolder));
+
+        Class<?> clazz = compiler.compiledClassOf(BASE_PACKAGE);
+        ActivationConfigProperty[] activationConfig = clazz.getAnnotation(MessageDriven.class).activationConfig();
+        assertThat(activationConfig, hasItemInArray(
+                allOf(propertyName(equalTo("destinationType")), propertyValue(equalTo("javax.jms.Topic")))));
```
Comment (matt-rich) [inline adapters/messaging-adapter-generator/src/test/java/uk/gov/justice/raml/jms/core/JmsEndpointGeneratorTest.java]: We should reuse the constants suggested above in the test case

Comment (jaceko) [inline adapters/messaging-adapter-generator/src/test/java/uk/gov/justice/raml/jms/core/JmsEndpointGeneratorTest.java]: that would make the test coupled to implementation, you normally don't wanna do it

File: adapters/messaging-adapter-generator/src/test/java/uk/gov/justice/raml/jms/core/JmsEndpointGeneratorTest.java [hunk 10]
```diff
@@ -547,6 +549,51 @@ public void shouldCallJmsProcessorWhenOnMessageIsInvoked() throws Exception {
...
+                        .with(resource()
+                                .withRelativeUri("/people.controller.commands")
+                                .with(action(POST, "application/vnd.people.events.abc+json")))
+                        .build(),
+                configurationWithBasePackage(BASE_PACKAGE, outputFolder));
+
+        Class<?> clazz = compiler.compiledClassOf(BASE_PACKAGE);
+        ActivationConfigProperty[] activationConfig = clazz.getAnnotation(MessageDriven.class).activationConfig();
+        assertThat(activationConfig, hasItemInArray(allOf(
+                propertyName(equalTo("destinationType")),
```
Comment (matt-rich) [inline adapters/messaging-adapter-generator/src/test/java/uk/gov/justice/raml/jms/core/JmsEndpointGeneratorTest.java]: Constants for String values

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/uri/ResourceUri.java (hunk 8 above)
Comment (jaceko) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/uri/ResourceUri.java]: nothing more meaningful than ONE, FOUR etc. don't see match value in it

File: adapters/messaging-adapter-generator/src/test/java/uk/gov/justice/raml/jms/core/JmsEndpointGeneratorTest.java (hunk 10 above)
Comment (jaceko) [inline adapters/messaging-adapter-generator/src/test/java/uk/gov/justice/raml/jms/core/JmsEndpointGeneratorTest.java]: tests should be verbose and descriptive, a form of documentation, you rarely use constants for stuff that is relevant to the test case, anything else goes to builder's defaults

File: adapters/rest-adapter-generator/src/test/java/uk/gov/justice/api/resource/DefaultUsersUserIdResourceIT.java [hunk 11]
```diff
@@ -200,7 +203,7 @@ public void shouldReturnResponseWithContentType() {
                 .path("/users/4444-5556")
                 .header("Accept", "application/vnd.people.queries.get-user+json")
                 .get();
-
+        assertThat(response.getStatus(), is(200));
```
Comment (matt-rich) [inline adapters/rest-adapter-generator/src/test/java/uk/gov/justice/api/resource/DefaultUsersUserIdResourceIT.java]: Bad 200 magic number, use Http Response Success Constant (and replace elsewhere also)

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/uri/ResourceUri.java (hunk 7 above)
Comment (jaceko) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/uri/ResourceUri.java]: will do it for both regular expressions when your refactoring goes through to the master, e.g. events->event, commands->command etc.
we currently don't support view queries via JMS, so not an issue for now

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/uri/BaseUri.java (hunk 6 above)
Comment (jaceko) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/uri/BaseUri.java]: will do it for both regular expressions when your refactoring goes through to the master, e.g. events->event, commands->command etc.
we currently don't support queries via JMS, so not an issue for now

File: adapters/messaging-adapter-generator/pom.xml (hunk 1 above)
Comment (jaceko) [inline adapters/messaging-adapter-generator/pom.xml]: done

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java (hunk 2 above)
Comment (jaceko) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java]: done

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java (hunk 3 above)
Comment (jaceko) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java]: done

File: adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java (hunk 5 above)
Comment (jaceko) [inline adapters/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/JmsEndpointGenerator.java]: done

**Pull request merged**